### DISCUSSION TITLE
Deploying via Thirdweb seems superior to Foundry-only. Thoughts?

### DISCUSSION THREAD

**Discussion by ewebgh33** [Cyfrin/foundry-full-course-cu#1415] [General]
Hi
As title says, to me (and so far) it seems as though deploying via ThirdWeb would be superior to a Foundry-only deployment?
Maybe I am missing something, I have finished lesson 6 so perhaps there are reasons that come up later.

For me, having a GUI to interact is vastly superior compared to needing to everything via CLI.
Even the additional step of having to convert from hex to dec every time you want to retrieve a value using Foundry? Who (at Foundry) though that was a useful idea... an extra step every time? vs TW lets you use token native amounts, like 0.1e not 10000000whatevever wei. Converion _built-in._

And then, simply just having _buttons_ (like Remix) in ThirdWeb is faster than typing out commands in Foundry to go check/update some values.

What am I missing here? Am I super wrong?
What are the down-sides to using Thirdweb (other than immediate verification, which you may not want, say if you're doing a stealth mint of a token and don't want to verify until just before mint).

Looking for input from the man himself @PatrickAlphaC or others who have ventured further than I have, and can advise.

Disclaimer: _Foundry course is still awesome, I am just saying the DEPLOY step seems to give me better access to more visual tools to interact more quickly, when I can then immediately use the ThirdWeb dashboard rather than continue in CLI. I am still very interested in dev and unit testing via Foundy._

Thanks!
E

**Comment by olahamid** [comment]
I am very curious to see someone else drop a good answer to this question, I'd love to learn from another POV.
for me I haven't used thirdweb at all so i can't really know how effective the GLI use case is compared to using CLI like foundry, but with foundry i know you get to simulate testing in a more manual approach hence you get a deeper understanding of what is happening behind the scene. also with foundry you get more control and customization to testing your smart contract using simulation like fuzzing to do extreme test. I am currently in lesson 12, i really dont have too much range of experience but i believe as you advance with patrick collin's foundry course you will see more use cases and get more answer to your question yourself.

**Reply by ewebgh33** [reply]
Testing in Foundry yes, but the deploy is still via CLI in VScode, just a different command.

On deploy you head over to the URL returned in your terminal. 

The URL is a dashboard and you get access to buttons to play with the contract. Because I also like to test by interacting on the web, like a user would, and this is a little closer to that. So not a replacement, just an additional thing you can do if you deploy with the ThirdWeb method.

**Comment by PatrickAlphaC** [comment]
Great questions! Let me try to address this. 

At the end of the day, there are often several tools out there to help you be most successful. Use the tools you like best! If you like thirdweb, great! Use it! 

> deploying via ThirdWeb would be superior to a Foundry-only deployment

Thirdweb is a great deployment option, and many people use it. It has advantages like Thirdweb deploy which allow for you to not have to input your private key! However, it has its downsides too. Using a GUI can be more restrictive for what you want to do. If the GUI doesn't support your desired functionality, you're a bit stuck! Additionally, a lot of developers want to have as little reliance on closed-sourced tooling as possible, because they can flip on you and turn into a paid service which can ruin your flow! 

Foundry's power comes more in the speed of its testing and developing, it's deploying leaves a lot to be desired. But on the backend, even thirdweb is doing something like what we are describing here. 

If you find deploying via Thirdweb more comfortable, I think that's a great option! 

> Even the additional step of having to convert from hex to dec ... Who ... though(t) that was a useful idea?

Most developers want to know exactly what the input/out was. Some numbers in some applications are _bad_ if you convert them, or more confusing. Some developers expect to get the raw response. So always converting can be actually _more_ confusing for developers. For starter developers, I agree, this can lead to some friction!

> simply just having buttons (like Remix) in ThirdWeb is faster

As your project grows, your deployment will often get more complicated, and you can write a script to handle everything in 1 command. However, in a GUI, you might have to press 50 buttons. That's really frustrating. Remix and Thirdweb are great tools for smaller projects, but the bigger the project gets, the more pressing a ton of buttons becomes much slower than running in a CLI. But, if you can wrap up a script into a one button press, like I said above, then thirdweb (or even remix) can be great options. Just be sure to write your tests!! 

> What am I missing here? Am I super wrong?

Nope! You can (and should) use the tools you like best! 

## Summary

You can (and should) use the tools that you like. If you like thirdweb, that's great! Every tool often comes with advantages and disadvantages and it's up to you to pick what's best for you:

Thirdweb advantages:
- Nice GUI
- Easier deployment
- Simpler Key management 

Cons:
- Less flexibility in deployment setup 
- Reliance on a centralized closed-sourced system for deployment

**Reply by ewebgh33** [reply]
Thanks so much for the answer. Everything you say (as usual) makes complete sense.
I am definitely not looking to get a job doing this - it's a side-gig for me, and mainly for my own enjoyment/entertainment. I have to be more mindful in future of that approach coloring my reactions to some things. :)
